src/SAWScript/SolverCache.hs

```diff
@@ -0,0 +1,632 @@
+  , optionBackends
+  , SolverBackendVersions
+  , SolverCacheKey(..)
+  , mkSolverCacheKey
+  , SolverCacheValue(..)
+  , toSolverCacheValue
+  , fromSolverCacheValue
+  , SolverCache(..)
+  , lazyOpenSolverCache
+  , openSolverCache
+  , SolverCacheOp
+  , lookupInSolverCache
+  , insertInSolverCache
+  , setSolverCachePath
+  , printSolverCacheByHex
+  , cleanSolverCache
+  , printSolverCacheStats
+  , testSolverCacheStats
+  ) where
+
+import System.Directory (createDirectoryIfMissing)
+import Control.Exception (try, SomeException)
+import Control.Monad (when, forM_)
+import System.Timeout (timeout)
+
+import GHC.Generics (Generic)
+import Data.IORef (IORef, newIORef, modifyIORef, readIORef)
+import Data.Time.Clock (UTCTime, getCurrentTime)
+import Data.Tuple.Extra (first, firstM)
+import Data.List (elemIndex, intercalate, isPrefixOf)
+import Data.Maybe (fromMaybe, isJust)
+import Data.Functor ((<&>))
+import Numeric (readHex)
+
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as M
+
+import qualified Data.Text as T
+import Data.Text.Encoding (encodeUtf8)
+import Text.Printf (printf)
+
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+
+import qualified Crypto.Hash.SHA256 as SHA256
+
+import Data.Aeson ( FromJSON(..), ToJSON(..), FromJSONKey(..), ToJSONKey(..)
+                  , (.:), (.:?), (.=), fromJSON )
+import qualified Data.Aeson as JSON
+import Codec.CBOR.JSON (encodeValue, decodeValue)
+import Codec.Serialise (Serialise(..))
+
+import qualified Database.LMDB.Simple as LMDB
+import qualified Database.LMDB.Simple.Extra as LMDB
+
+import qualified Data.SBV.Dynamic as SBV
+
+import Verifier.SAW.FiniteValue
+import Verifier.SAW.SATQuery
+import Verifier.SAW.ExternalFormat
+import Verifier.SAW.SharedTerm
+
+import SAWScript.Options
+import SAWScript.Proof
+
+
+-- Orphan Instances and Helper Functions ---------------------------------------
+
+deriving instance Generic FirstOrderType
+deriving instance Generic FirstOrderValue
+
+-- | The options for JSON-serializing 'FirstOrderType's and 'FirstOrderValue's:
+-- remove the @FOV@/@FOT@ prefixes and encode the different constructors as
+-- two-element arrays.
+firstOrderJSONOptions :: JSON.Options
+firstOrderJSONOptions =
+  JSON.defaultOptions { JSON.sumEncoding = JSON.TwoElemArray
+                      , JSON.constructorTagModifier = dropFO }
+  where dropFO ('F':'O':tv:cs) | tv `elem` ['T', 'V'] = cs
+        dropFO cs = cs
+
+instance FromJSON FirstOrderType where
+  parseJSON = JSON.genericParseJSON firstOrderJSONOptions
+instance FromJSON FirstOrderValue where
+  parseJSON = JSON.genericParseJSON firstOrderJSONOptions
+
+instance ToJSON FirstOrderType where
+  toJSON = JSON.genericToJSON firstOrderJSONOptions
+  toEncoding = JSON.genericToEncoding firstOrderJSONOptions
+instance ToJSON FirstOrderValue where
+  toJSON = JSON.genericToJSON firstOrderJSONOptions
+  toEncoding = JSON.genericToEncoding firstOrderJSONOptions
```

Comment:
I generally agree with you, but in this case the fact that I used [`firstOrderJSONOptions`](https://github.com/GaloisInc/saw-script/blob/9c5c3949c92d5fb6115bf1760ac2cf4eff3c4bf8/src/SAWScript/SolverCache.hs#L132) instead of `defaultJSONOptions` means it's not quite as brittle as it could be. By default, each constructor would be encoded by an integer, thus making it brittle with respect to constructor reordering and new constructors being added in the middle of others. However, `firstOrderJSONOptions` specifically encodes constructors like `FOWord` as `"Word"`. This means this encoding is independent of constructor order and new constructors being added. As long as no constructor names are changed, out-of-date encodings should always work.

As for the `saw-script` `git` commit, that is only included in the case of `rme` being the backend, since that sub-component has no versioning system unto itself. So, if a constructor name is ever changed, it is possible that `saw-script` will attempt to use an out-of-date encoding when retrieving a non-`rme` cached entry.

Regardless, I definitely agree that this assumption (that constructor names will not change) should be documented. If your comment about moving instances to `FiniteValue.hs` include these `ToJSON`/`FromJSON` instances as well, I will put the comment in `FiniteValue.hs` as well.

Reply:
Added some docs (and resolved the comment above) in a960c3665db04562d30b9160682bcef359cb5fff!